### PR TITLE
Unix fsync

### PR DESCRIPTION
**Pull request by UnixJunkie**

### CODE REVIEW

**Comment by UnixJunkie** [comment]
related to https://caml.inria.fr/mantis/view.php?id=2533

**Comment by UnixJunkie** [comment]
I have no idea on how to do the windows part.
I hope a good will fill that bit.

File: otherlibs/win32unix/unix.ml [hunk 1]
```diff
@@ -182,6 +182,8 @@ type file_perm = int
 external openfile : string -> open_flag list -> file_perm -> file_descr
            = "unix_open"
 external close : file_descr -> unit = "unix_close"
+external fsync : file_descr -> unit =
+  invalid_arg "Unix.fsync not implemented"
```
Comment (yallop) [inline otherlibs/win32unix/unix.ml]: `external` ↝ `let`

**Comment by nojb** [comment]
For Windows, you should use [`FlushFileBuffers`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364439%28v=vs.85%29.aspx), see e.g. http://ab-initio.mit.edu/octave-Faddeeva/gnulib/lib/fsync.c

File: otherlibs/win32unix/unix.ml (hunk 1 above)
Comment (dbuenzli) [reply otherlibs/win32unix/unix.ml]: Isn't there a compatibility equivalent on windows ? If not please indicate that in the API docs and in the manual table listing platform differences in Unix that  that `fsync` is not implemented on windows.

**Comment by UnixJunkie** [comment]
added, thanks for the link, and thanks to @rwmjones for the code?

File: otherlibs/win32unix/unix.ml [hunk 2]
```diff
@@ -181,7 +181,9 @@ type file_perm = int
 
 external openfile : string -> open_flag list -> file_perm -> file_descr
            = "unix_open"
-external close : file_descr -> unit = "unix_close"
+external close : file_descr -> unit = "fsync"
+external fsync : file_descr -> unit =
+  invalid_arg "Unix.fsync not implemented"
```
Comment (dbuenzli) [inline otherlibs/win32unix/unix.ml]: ?????

File: otherlibs/win32unix/fsync.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+   http://www.sqlite.org/copyright.html
+
+   Written by Richard W.M. Jones <rjones.at.redhat.com>
+
+   Copyright (C) 2008-2012 Free Software Foundation, Inc.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
```
Comment (dbuenzli) [inline otherlibs/win32unix/fsync.c]: This may not be compatible with the compiler's license (has no static linking exception).

**Comment by nojb** [comment]
The link was just for inspiration, to see how the function is called; you still need to write a stub function under Windows, with signature `CAMLprim value unix_fsync(value fd)`.

**Comment by UnixJunkie** [comment]
I don't have access to a windows computer, if you know better, feel free to improve the PR.

File: otherlibs/win32unix/fsync.c (hunk 3 above)
Comment (dra27) [reply otherlibs/win32unix/fsync.c]: I can make that stronger - this code is *not* compatible with the compiler's contribution requirements. See https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#contributor-license-agreement. @rwmjones may be willing to contribute it to the compiler sources.

**Comment by dra27** [comment]
@UnixJunkie - contributions are always welcome, so please don't take this in the wrong way. Contributions which move one platform forwards while leaving another behind necessarily create work and, in my opinion at least, are better worked on elsewhere - Discuss, caml-list and so forth can be used to draw developers in to help.

The Windows version of fsync.c will need adapting from that version you've used, as that's only been designed for mingw (i.e. GCC). There are various pointers which can get around your lack of a Windows machine:

 - I have [blogged](http://www.dra27.uk/blog/platform/2017/09/27/ocaml-github-ci.html) on setting up AppVeyor on one's own OCaml fork.
 - https://github.com/ocaml/opam/blob/master/appveyor.yml#L43-L46 contains a sample of how to get AppVeyor to pause at the end of a build (failed or otherwise) which allows you to connect to the build machine using an RDP client. **Do not use this setting on OCaml's AppVeyor - please set it up on your own**.
 - On a private branch, you can of course disable parts of the AppVeyor build so that you get to the bit of the build you want as quickly as possible (although msvc64 runs early in the process anyhow).
 - Surely your university provides Microsoft Imagine, and therefore easy and free access to a Windows 10 Licence which you could virtualise?

**Comment by nojb** [comment]
I would like to emphasize @dra27's message. OCaml takes Windows compatibility seriously, and any improvement should ideally apply to all platforms (if at all possible).

With regards to this PR, I went ahead and pushed a commit with a possible implementation of `fsync` for Windows (FWIW, it is written from scratch without using any existing code).

**Comment by dra27** [comment]
Reading up on MSDN's entry for [`FlushFileBuffers`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364439(v=vs.85).aspx), I'm slightly concerned at the behaviour mentioned for pipes. [`fsync(2)`](http://man7.org/linux/man-pages/man2/fsync.2.html) should return `EINVAL` for a pipe, but according to MSDN, `FlushFileBuffers` will block until the pipe has been fully read.

It's belt-and-braces, as we don't support named pipes in the Unix library itself, but should the call be guarded with a call to [`GetFileType`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364960(v=vs.85).aspx)?

**Comment by nojb** [comment]
I don't have a strong opinion either way - I will add the check.

**Comment by pmetzger** [comment]
Side comment: I didn't even realize we lacked `fsync`, which is Very Very Important at times.

File: otherlibs/unix/fsync.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           */
```
Comment (dra27) [inline otherlibs/unix/fsync.c]: This copyright header is wrong!

File: otherlibs/unix/unix.mli [hunk 5]
```diff
@@ -307,6 +307,9 @@ val openfile : string -> open_flag list -> file_perm -> file_descr
 val close : file_descr -> unit
 (** Close a file descriptor. *)
 
+val fsync : file_descr -> unit
+(** Synchronize a file's in-core state with the storage device. *)
```
Comment (dra27) [inline otherlibs/unix/unix.mli]: Does `fsync` ever cause anything to be read? Synchronize to me implies both directions: how about "Flushes all in-core data of the file associated with the `file_descr` to the storage device."?

File: otherlibs/win32unix/Makefile [hunk 6]
```diff
@@ -34,7 +34,7 @@ UNIX_FILES = access.c addrofstr.c chdir.c chmod.c cst2constr.c \
   exit.c getaddrinfo.c getcwd.c gethost.c gethostname.c \
   getnameinfo.c getproto.c \
   getserv.c gmtime.c mmap_ba.c putenv.c rmdir.c \
-  socketaddr.c strofaddr.c time.c unlink.c
+  socketaddr.c strofaddr.c time.c unlink.c fsync.c
```
Comment (dra27) [inline otherlibs/win32unix/Makefile:37]: This shouldn't be here.

**Review by dra27 — COMMENTED** [review]
A few minor tweaks - and it needs a `Changes` entry.

File: otherlibs/unix/unix.mli (hunk 5 above)
Comment (dra27) [reply otherlibs/unix/unix.mli]: (I realise the man page for `fsync` begins by using the word synchronisation - it's just that it then goes on to describe flushing only AFAICT! FlushFileBuffers definitely does only what it says on the tin...)

File: otherlibs/unix/fsync.c (hunk 4 above)
Comment (UnixJunkie) [reply otherlibs/unix/fsync.c]: Where can I get a correct one then?

File: otherlibs/win32unix/Makefile (hunk 6 above)
Comment (UnixJunkie) [reply otherlibs/win32unix/Makefile:37]: where should it be then?

File: otherlibs/unix/unix.mli (hunk 5 above)
Comment (pmetzger) [reply otherlibs/unix/unix.mli]: The phrasing was taken straight from the summary line of the man page in *BSD and OS X.

I agree, though, that "flush in-memory file buffers to disk" or some such is better.

File: otherlibs/win32unix/Makefile (hunk 6 above)
Comment (dra27) [reply otherlibs/win32unix/Makefile:37]: Deleted. Files in this list are intended to be copied from the Unix version of the library.

File: otherlibs/unix/fsync.c (hunk 4 above)
Comment (dra27) [reply otherlibs/unix/fsync.c]: Your name isn’t Xavier, you’re not at Inria, and I hope you haven’t been sitting on the patch for 22 years? 🙂

File: otherlibs/win32unix/Makefile (hunk 6 above)
Comment (UnixJunkie) [reply otherlibs/win32unix/Makefile:37]: edits from maintainers are allowed...

**Comment by UnixJunkie** [comment]
Isn't _commit the equivalent of fsync on windows?

**Comment by UnixJunkie** [comment]
Maybe good to go.
I put @nojb as the first contributor, since he did the most work.

**Comment by dra27** [comment]
You learn a new thing every morning - it does indeed look as though `_commit` is equivalent to `fsync` (a quick look online suggest Postgres just `#ifdef`s it)

The Windows code here is doing the same thing - @nojb, the code could be shared via `fsync_os`. What do you think?

**Comment by nojb** [comment]
Yes, it looks equivalent to what we are doing - but indeed it would be better since we could share the Unix/Windows code. Will amend.

**Comment by nojb** [comment]
I made a commit replacing `FlushFileBuffers` with `_commit`; @dra27 do you mind taking a look?

**Comment by damiendoligez** [comment]
I'm not sure, but we might need a CLA for this contribution. (cc @xavierleroy)

**Comment by UnixJunkie** [comment]
doesn't the windows code needs
```
#include <io.h>
```
just under 
```
#ifdef _WIN32 
```
?

**Comment by nojb** [comment]
> doesn't the windows code needs

Since the CI passes, I guess it gets included via some other header, but anyway I went ahead and added it for extra safety.

**Comment by UnixJunkie** [comment]
I think this PR is ready to merge.
Thanks to all those who contributed!

**Comment by dra27** [comment]
LGTM - the CLA question needs resolving before it can be squashed. You have left the copyright on `fsync.c` as being with Inria, which I presume you don't care about.

**Comment by dra27** [comment]
The `<io.h>` thing is bothering me (mainly as to why it was working before - I can't see how that header was being included, at least just looking at the code).

**Comment by fdopen** [comment]
mingw includes `io.h` through `direct.h` (included in `unixsupport.h`). msvc probably won't notify you, because the warning level is too low and `_commit` returns `int` (default return type).

**Comment by dra27** [comment]
Thanks @fdopen!

**Comment by UnixJunkie** [comment]
Yes, I wanted to leave the copyright to INRIA.

**Comment by UnixJunkie** [comment]
@gasche I think it is ready to merge.

**Comment by gasche** [comment]
I haven't participated in this discussion nor reviewed the code, I'll let other maintainers who did make decisions.

**Comment by nojb** [comment]
Merged on the basis of reviews by myself and @dra27.

**Comment by UnixJunkie** [comment]
thanks!

File: otherlibs/unix/fsync.c [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+/*                                                                        */
+/**************************************************************************/
+
+#include <caml/mlvalues.h>
+#include <caml/signals.h>
+#include "unixsupport.h"
+
+#ifdef _WIN32
+#include <io.h>
+#define fsync(fd) _commit(win_CRT_fd_of_filedescr(fd))
```
Comment (ygrek) [inline otherlibs/unix/fsync.c:22]: is this safe? fd is custom block and cannot be accessed outside of blocking section

Comment (nojb) [reply otherlibs/unix/fsync.c:22]: Good catch, thanks! Fix in https://github.com/ocaml/ocaml/pull/1949.

**Pull request merged**